LibJS: Module support

davidot opened the pull request:
@Lubrsi I used part of your environment setting PR and gave you authorship of that commit (and stole the message) with updated copyrights where I saw them. Let me know if this should be done differently.

TODO:
- [x] Implement all types of exports and imports (I think)
- [x] Figure out what it means to execute a module
- [x] Test/create libjs-test262 patch https://github.com/linusg/libjs-test262/pull/50
- [x] Run test262
- [x] Add import.meta support
- [x] Add top level await support
- [x] Add local tests 
- [x] Test all other parts of the system (Already did it with the refactor interpreter patch)
- [x] Disable module loading in the browser for now? Or at least some host hook since this can now execute arbitrary local Javascript from a website I think.

:warning: Must be merged at the same time as https://github.com/linusg/libjs-test262/pull/50 for test262 not to fail :warning:

davidot commented:
This is now at least ready for review, I will clean up the commits and fixme later in the week however I'm a bit tired of modules for the moment.

I will leave some comments about questionable things which I would like comments on.

davidot commented on Userland/Libraries/LibJS/Module.h:
```diff
@@ -22,15 +64,34 @@ class Module : public RefCounted<Module> {
+    }
+
 private:
-    // Handles are not safe unless we keep the VM alive.
-    NonnullRefPtr<VM> m_vm;
```
This `RefPtr<VM>` was causing cycles since we (have to) store Modules in the VM also by RefPTR.
Instead we simply take a VM parameter if you want to do anything useful with a Module.
This might become a problem later with the browser but I think we can leave that for another time. 
Also we can always store a reference in [[CustomData]] so it should be extensible enough to work.

davidot commented on Userland/Libraries/LibJS/Tests/builtins/ShadowRealm/ShadowRealm.prototype.importValue.js:
```diff
@@ -3,18 +3,64 @@ describe("normal behavior", () => {
+        expect(error.message).toBe("Could not find module: './file_should_not_exists.not.js'");
+    });
+
+    test("basic functionality", () => {
+        const shadowRealm = new ShadowRealm();
+        const promise = shadowRealm.importValue("./external-module.mjs", "foo");
+        expect(promise).toBeInstanceOf(Promise);
+        let error = null;
+        let passed = false;
+        promise
+            .then(value => {
+                expect(value).toBe("Well hello shadows");
+                expect(typeof value).toBe("string");
+
+                expect(value).not.toHaveProperty("default", null);
+                expect(value).not.toHaveProperty("bar", null);
+                passed = true;
+            })
+            .catch(value => {
+                error = value;
+            });
+        runQueuedPromiseJobs();
+        expect(error).toBeNull();
+        expect(passed).toBeTrue();
+    });
```
I just copied the original tests but I have no idea what I should tests here.

linusg replied:
Looks fine!

davidot commented:
For completeness sake here is an example of the browser rejecting a dynamic import

![browser_dynamic_import_failure](https://user-images.githubusercontent.com/3750649/150209507-111a98c2-27f1-4534-9269-f0019caa7905.png)

linusg commented on Userland/Libraries/LibJS/SourceTextModule.cpp:
```diff
@@ -19,18 +81,571 @@ Result<NonnullRefPtr<SourceTextModule>, Vector<Parser::Error>> SourceTextModule:
+        }
+
+        for (auto const& export_entry : export_statement.entries()) {
+            // FIXME: Store default entry
```
Could also be a bit clearer IMO.

davidot replied:
This FIXME was already resolved just didn't remove the comment

linusg commented on Userland/Libraries/LibJS/SourceTextModule.cpp:
```diff
@@ -19,18 +81,571 @@ Result<NonnullRefPtr<SourceTextModule>, Vector<Parser::Error>> SourceTextModule:
+        auto resolution = TRY(resolve_export(vm, entry.export_name));
+        // b. If resolution is null or ambiguous, throw a SyntaxError exception.
+        if (!resolution.is_valid())
+            return throw_completion(SyntaxError::create(realm().global_object(), String::formatted("Invalid or ambiguous export entry '{}'", entry.export_name)));
```
Can we add an `ErrorType` for this?

linusg commented on Userland/Libraries/LibJS/SourceTextModule.cpp:
```diff
@@ -19,18 +81,571 @@ Result<NonnullRefPtr<SourceTextModule>, Vector<Parser::Error>> SourceTextModule:
+    }
+
+    // 2. Assert: All named exports from module are resolvable.
+    // FIXME: Isn't this what step 1 does?
```
Yeah, I think this is just to assert that *at this point* all of them are resolvable, because we threw an exception otherwise.

linusg replied:
(FIXME might not be needed)

davidot replied:
Yeah I think they are also talking about the local export entries perhaps.
I have replaced the FIXME with a comment.
Although I'm still not a 100% that this is what they mean.

linusg commented on Userland/Libraries/LibJS/SourceTextModule.cpp:
```diff
@@ -19,18 +81,571 @@ Result<NonnullRefPtr<SourceTextModule>, Vector<Parser::Error>> SourceTextModule:
+
+    // 3. Let realm be module.[[Realm]].
+    // 4. Assert: realm is not undefined.
+    VERIFY(&realm());
```
This seems a bit silly, since `realm()` is already guaranteed to be a (non-null) reference :^)

linusg commented on Userland/Libraries/LibJS/SourceTextModule.cpp:
```diff
@@ -19,18 +81,571 @@ Result<NonnullRefPtr<SourceTextModule>, Vector<Parser::Error>> SourceTextModule:
+    // 6. Set module.[[Environment]] to env.
+    set_environment(environment);
+
+    auto& global_object = realm().global_object();
```
We could do this earlier an also use it for the `ModuleEnvironment` above

linusg commented on Userland/Libraries/LibJS/SourceTextModule.cpp:
```diff
@@ -19,18 +81,571 @@ Result<NonnullRefPtr<SourceTextModule>, Vector<Parser::Error>> SourceTextModule:
+    // 7. For each ImportEntry Record in of module.[[ImportEntries]], do
+    for (auto& import_entry : m_import_entries) {
+        if (!import_entry.module_request().assertions.is_empty())
+            return throw_completion(InternalError::create(global_object, String::formatted(ErrorType::NotImplemented.message(), "import statements with assertions")));
```
Could be simplified as `return vm.throw_completion<InternalError>(global_object, ErrorType::NotImplemented, "import statements with assertions")`

davidot replied:
I intentionally don't use `vm.throw_...` in this code because it sets exception and this means I never have to clear it.

Should I switch to that? Or should we perhaps add a convenience function to create an exception like this without setting it on the VM?

linusg commented on Userland/Libraries/LibJS/SourceTextModule.cpp:
```diff
@@ -19,18 +81,571 @@ Result<NonnullRefPtr<SourceTextModule>, Vector<Parser::Error>> SourceTextModule:
+
+            // ii. If resolution is null or ambiguous, throw a SyntaxError exception.
+            if (!resolution.is_valid())
+                return throw_completion(SyntaxError::create(realm().global_object(), String::formatted("Invalid or ambiguous export entry '{}'", import_entry.import_name)));
```
Same as above, `ErrorType` + can use template variant of `throw_completion()`

linusg commented on Userland/Libraries/LibJS/SourceTextModule.cpp:
```diff
@@ -19,18 +81,571 @@ Result<NonnullRefPtr<SourceTextModule>, Vector<Parser::Error>> SourceTextModule:
+    // 9. Set the Function of moduleContext to null.
+
+    // 10. Assert: module.[[Realm]] is not undefined.
+    VERIFY(&realm());
```
Again here :)

linusg commented on Userland/Libraries/LibJS/SourceTextModule.cpp:
```diff
@@ -19,18 +81,571 @@ Result<NonnullRefPtr<SourceTextModule>, Vector<Parser::Error>> SourceTextModule:
+    async_block_start(vm, m_ecmascript_code, capability.value(), module_context);
+
+    // c. Return NormalCompletion(empty).
+    return js_undefined();
```
I'm afraid this function will need to return a proper Completion to represent this return state - undefined is not the same as empty.

linusg commented on Userland/Libraries/LibJS/Interpreter.cpp:
```diff
@@ -130,11 +130,31 @@ ThrowCompletionOr<Value> Interpreter::run(Script& script_record)
+        auto* error = vm.heap().allocate<Exception>(global_object(), *(evaluated_or_error.throw_completion().value()));
+        vm.set_exception(*error);
+        return evaluated_or_error.throw_completion();
+    }
```
Since this API is brand new and not being used anywhere yet, do we actually need to do this? Only checking for a throw completion at the call site should be fine.

davidot replied:
Although it is not directly visible in the commit this is used by `js` with 
```cpp
result = script_or_module.visit(
  [&](auto& visitor) {
      return interpreter.run(*visitor);
  });
```

And `js` currently breaks if you return an error but don't set exception.
(Although I guess we could just set the exception in `js` I'll try that real quick)

davidot replied:
Ended up fixing this in `js` which is nicer :+1:

linusg commented on Userland/Libraries/LibJS/Runtime/VM.cpp:
```diff
@@ -682,4 +689,171 @@ ScriptOrModule VM::get_active_script_or_module() const
+        if (stored_module.filepath != filepath)
+            return false;
+        return true;
+        //        return stored_module.referencing_script_or_module.visit(
+        //            [&](Empty) {
+        //                return script_or_module.has<Empty>();
+        //            },
+        //            [&](Script* script) {
+        //                return script_or_module.has<Script*>() && script_or_module.get<Script*>() == script;
+        //            },
+        //            [&](Module* module) {
+        //                return script_or_module.has<Module*>() && script_or_module.get<Module*>() == module;
+        //            });
```
Commented out code - left from debugging?

davidot replied:
This is a bit tricky because the spec says:
`Each time this operation is called with a specific referencingScriptOrModule, specifier pair as arguments it must return the same Module Record instance if it completes normally.`

So perhaps we need to search with both at some point. But for now this works so I have added a comment.

linusg commented on Userland/Libraries/LibJS/Runtime/VM.cpp:
```diff
@@ -682,4 +689,171 @@ ScriptOrModule VM::get_active_script_or_module() const
+    VERIFY(!exception());
+
+    if (evaluated_or_error.is_error())
+        return JS::throw_completion(*evaluated_or_error.throw_completion().value());
```
Can we not return the throw completion directly instead of wrapping it?

linusg commented on Userland/Libraries/LibJS/Runtime/VM.cpp:
```diff
@@ -682,4 +689,171 @@ ScriptOrModule VM::get_active_script_or_module() const
+    auto file_or_error = Core::File::open(filepath, Core::OpenMode::ReadOnly);
+
+    if (file_or_error.is_error()) {
+        return JS::throw_completion(Error::create(global_object, String::formatted("Could not find module: '{}'", specifier.module_specifier)));
```
`ErrorType` + templated `throw_completion`

linusg commented on Userland/Libraries/LibJS/AST.cpp:
```diff
@@ -4097,6 +4171,12 @@ void ExportStatement::dump(int indent) const
+
+    if (m_statement) {
+        print_indent(indent + 1);
+        outln("(Expression)");
```
Perhaps `(Statement)` to match the member?

linusg commented on Userland/Libraries/LibJS/Runtime/VM.cpp:
```diff
@@ -50,6 +50,37 @@ VM::VM(OwnPtr<CustomData> custom_data)
+        // If you are here because you want to enable dynamic module importing make sure it won't be a security problem
+        // by checking the default implementation of HostImportModuleDynamically and creating your own hook or calling
+        // vm.enable_default_host_import_module_dynamically_hook().
+        promise->reject(Error::create(global_object, String::formatted("Dynamic Imports are not allowed")));
```
`ErrorType` for this as well please

linusg commented on Userland/Libraries/LibJS/Runtime/VM.cpp:
```diff
@@ -856,4 +894,111 @@ ThrowCompletionOr<NonnullRefPtr<Module>> VM::resolve_imported_module(ScriptOrMod
+
+    // FIXME: I don't understand why we have this internal promise? Maybe we can just use a throw completion for this?
+    //        Maybe it is just the way the spec handles doing this (potentially) asynchronously
+    Promise* promise = Promise::create(global_object);
```
Could use `auto*` here

linusg commented on Userland/Libraries/LibJS/Runtime/VM.cpp:
```diff
@@ -856,4 +894,111 @@ ThrowCompletionOr<NonnullRefPtr<Module>> VM::resolve_imported_module(ScriptOrMod
+
+    VERIFY(!exception());
+
+    // Note: We handle these immediately as we don't have asynchronous processing yet.
+    run_queued_promise_jobs();
```
Is this actually needed? `Interpreter::run()` / the host should do it for us.

davidot replied:
I thought we needed this one but that was because I used references in the lambdas :facepalm:

linusg commented on Userland/Libraries/LibJS/Tests/modules/basic-modules.js:
```diff
@@ -0,0 +1,147 @@
+        .then(result => {
+            moduleLoaded = true;
+            moduleResult = result;
+            // expect(result).toHaveProperty('passed', true);
```
Commented out code

linusg commented on Userland/Libraries/LibJS/Tests/modules/basic-modules.js:
```diff
@@ -0,0 +1,147 @@
+
+    if (thrownError) {
+        throw thrownError;
+        // const err = new ExpectationError("Failed to import module: " + filename + " with error " + thrownError);
+        // err.cause = thrownError;
+        // throw err;
```
Commented out code

linusg commented on Userland/Libraries/LibJS/Tests/builtins/ShadowRealm/ShadowRealm.prototype.importValue.js:
```diff
@@ -3,18 +3,64 @@ describe("normal behavior", () => {
         // NOTE: The actual import is currently not implemented and always pretends to fail for now.
         const shadowRealm = new ShadowRealm();
-        const promise = shadowRealm.importValue("./myModule.js", "foo");
+        const promise = shadowRealm.importValue("./file_should_not_exists.not.js", "foo");
```
Typo: `should not exists`

linusg reviewed (commented):
Really great work, thanks for working on this :tada: 

Bunch of questions / suggestions, let me know if anything is unclear.

linusg reviewed (commented):
Also a small typo in `LibJS: Implement ImportCall and HostImportModuleDynamically`: `form a spec` -> `from a spec`

davidot commented on Userland/Libraries/LibJS/CyclicModule.h:
```diff
@@ -0,0 +1,63 @@
+    ThrowCompletionOr<void> async_module_execution_rejected(VM& vm, Value error);
+
+    ModuleStatus m_status { ModuleStatus::Unlinked };   // [[Status]]
+    Completion m_evaluation_error;                      // [[EvaluationError]] FIXME: Should this be an Optional<Completion> ?
```
Yeah that's much nicer!

davidot commented on Userland/Libraries/LibJS/CyclicModule.h:
```diff
@@ -0,0 +1,63 @@
+    Optional<u32> m_dfs_ancestor_index;                 // [[DFSAncestorIndex]]
+    Vector<FlyString> m_requested_modules;              // [[RequestedModules]]
+    CyclicModule* m_cycle_root;                         // [[CycleRoot]]
+    bool m_has_TLA { false };                           // [[HasTLA]]
```
Yeah I wanted to clarify that this was an acronym but I've now changed it to: `has_top_level_await` so you don't have to know what it stands for.

davidot commented on Userland/Libraries/LibJS/Module.h:
```diff
@@ -11,6 +12,47 @@
 namespace JS {
 
+struct ResolvedBinding {
+    // FIXME: This isn't the nicest way to do this I think...
```
I looked at it again and I'm actually fairly ok with the current implementations.
I think this was originally about using `Variant` to support the spec way of saying: `[[BindingName]]: String | namespace`.

So unless you have some comments about this I will remove the FIXME.

davidot commented:
I think I addressed all the comments and the typo in the commit message, I have left a couple of discussions open as there might still be something to discuss there.

davidot commented on Userland/Libraries/LibJS/Module.cpp:
```diff
@@ -1,22 +1,120 @@
+// 16.2.1.5.1.1 InnerModuleLinking ( module, stack, index ), https://tc39.es/ecma262/#sec-InnerModuleLinking
+ThrowCompletionOr<u32> Module::inner_module_linking(VM& vm, Vector<Module*>&, u32 index)
+{
+    // Note: Until we have something extending module which is not SourceTextModule we crash.
```
> Could be a TODO
This (and the one below) are already implemented in SourceTextModule but have to be implemented in the spec for Modules
However currently SourceTextModule is the only non-abstract class so if we end up in this method we did something wrong.

Should I maybe change the comment to be more clear about that?

linusg commented on Userland/Libraries/LibJS/Runtime/ModuleNamespaceObject.cpp:
```diff
@@ -0,0 +1,197 @@
+void ModuleNamespaceObject::initialize(GlobalObject& global_object)
+{
+    Object::initialize(global_object);
+    define_direct_property(*vm().well_known_symbol_to_string_tag(), js_string(vm(), "Module"sv), PropertyAttributes(0));
```
```suggestion

    // 28.3.1 @@toStringTag, https://tc39.es/ecma262/#sec-@@tostringtag
    define_direct_property(*vm().well_known_symbol_to_string_tag(), js_string(vm(), "Module"sv), 0);
```

(yes, that's the actual title and URL :thonk:)

davidot replied:
The only true `@@toStringTag`? That's weird

linusg commented on Userland/Libraries/LibJS/Runtime/ModuleNamespaceObject.cpp:
```diff
@@ -0,0 +1,197 @@
+    auto value = TRY(internal_get(property_key, this));
+
+    // 5. Return PropertyDescriptor { [[Value]]: value, [[Writable]]: true, [[Enumerable]]: true, [[Configurable]]: false }.
+    return PropertyDescriptor { value, nullptr, nullptr, true, true, false };
```
```suggestion
    return PropertyDescriptor { .value = value, .writable = true, .enumerable = true, .configurable = false };
```

emanuele6 commented on Userland/Libraries/LibJS/CyclicModule.cpp:
```diff
@@ -0,0 +1,658 @@
+    (void)TRY(initialize_environment(vm));
+
+    // 11. Assert: module occurs exactly once in stack.
+    auto count = 0;
+    for (auto* module : stack) {
+        if (module == this)
+            count++;
+    }
+    VERIFY(count == 1);
```
I notice this while watching @linusg's video.

Wouldn't it be more correct to use `size_t` here?

`stack`  is an `AK::Vector` that can have up to `SIZE_MAX` elements.
Here `count` would be a `signed int`.

Also the only purpose of this code is to assert that `count == 1` so it is probably not great that a `SIZE_MAX` (on 64 bit) can overflow a `signed int` 4_294_967_295 times (maybe my math is slightly wrong, but somewhere around that number :D).

emanuele6 replied:
merged 13 seconds ago?? man, my timing sucks!
